atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java

```diff
@@ -218,11 +228,47 @@ protected SnapshotTransaction createTransaction(
         return runTaskThrowOnConflict(task, new ReadTransaction(transaction, sweepStrategyManager));
     }
 
+    /**
+     * Registers a Runnable that will be run when the transaction manager is closed, provided no callback already
+     * submitted throws an exception.
+     *
+     * Concurrency: If this method races with close(), then closingCallback may not be called.
+     */
+    public void registerClosingCallback(Runnable closingCallback) {
+        Preconditions.checkNotNull(closingCallback, "Cannot register a null callback.");
+        closingCallbacks.add(closingCallback);
+    }
+
+    /**
+     * Frees resources used by this SnapshotTransactionManager, and invokes any callbacks registered to run on close.
+     * This includes the cleaner, the key value service (and attendant thread pools), and possibly the lock service.
+     *
+     * Concurrency: If this method races with registerClosingCallback(closingCallback), then closingCallback
+     * may be called (but is not necessarily called). Callbacks registered before the invocation of close() are
+     * guaranteed to be executed (because we use a synchronized list) as long as no exceptions arise. If an exception
+     * arises, then no guarantees are made with regard to subsequent callbacks being executed.
+     */
     @Override
     public void close() {
-        super.close();
-        cleaner.close();
-        keyValueService.close();
+        if (isClosed.compareAndSet(false, true)) {
+            super.close();
+            cleaner.close();
+            keyValueService.close();
+            closeLockServiceIfPossible();
+            for (Runnable callback : closingCallbacks) {
+                callback.run();
+            }
+        }
+    }
+
+    private void closeLockServiceIfPossible() {
+        if (lockService instanceof Closeable) {
```

Comment:
Yep, makes sense. As far as I know all actual implementations of the lock service use regular `Closeable`, but no reason we should fail to close something that's `AutoCloseable` but not `Closeable`